Controllers films users

MarchenkoDmitrii opened the pull request.

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
no idea

MarchenkoDmitrii committed:
with idea

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

MarchenkoDmitrii committed:
first commit

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,54 @@
+@RestController
+public class FilmController {
+    private int id = 0;
+    Map<Integer, Film> films = new HashMap<>();
```
Добавь модификатор доступа

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,54 @@
+    Map<Integer, Film> films = new HashMap<>();
+    private static final Logger log = LoggerFactory.getLogger(FilmController.class);
+
+    @GetMapping("/films")
```
Вот это значение в скобках можно установить к аннотации @requestmapping над контроллером и не повторять это над каждым эндпоинтом

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,55 @@
+@RestController
+public class UserController {
+    protected int id = 0;
+    Map<Integer, User> users = new HashMap<>();
```
Модификатор нужен

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.time.LocalDate;
+
+@lombok.Data
```
Здесь импорт можно поправить, оставив только @Data

PolnySkvorets reviewed (changes requested):
Посмотри замечания. В целом, окей, но по мелочам пропустил

MarchenkoDmitrii committed:
1. Added access modification for "films" and "users" hashmaps
2. Added @RequestMapping annotation for filmcontroller and usercontroller
3. Changed import @Data lombook for models

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ValidationException.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.http.HttpStatus;
+
+public class ValidationException extends IllegalArgumentException {
+    public ValidationException(String cause, HttpStatus status) {
```
Вот здесь  status не используется никак, может его убрать или добавить логику?

MarchenkoDmitrii replied:
В конструкторе она принимается, разве нет? Логика реализована в других классах для возможности использовать одно исключение во всех возможных случаях, например: 
``
  throw new ValidationException("логин не может быть пустым и содержать пробелы", HttpStatus.BAD_REQUEST);
  ``
Тут выводится причина прерывания и передается http статус

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @GetMapping
+    public List<Film> findAll() {
+        if (films.equals(null))
```
это условие никогда не выполнится, тем более даже если оно выполнится - упадёт  NPE, чтобы этого избежать, нужно объекты сравнивать безопасно, например Objects.equals(films, null)

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @PostMapping
+    public Film create(@RequestBody Film film) {
+        if (film.getName().equals("") || film.getName().isEmpty())
```
нет проверки на  null, . можно поймать NPE

PolnySkvorets reviewed (changes requested):
посмотри замечания по валидации

MarchenkoDmitrii committed:
1. Исправлена проверка отсутствия фильмов;
2. В Эндпойнте @POSTMapping добавленна проверка на null.

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -22,13 +19,15 @@ public class FilmController {
-        if (films.equals(null))
+        if (films.entrySet().size() == 0)
             throw new ValidationException("Нет фильмов", HttpStatus.NO_CONTENT);
-        return films.values().stream().collect(Collectors.toList());
```
Да, так лучше, стримы нужны для фильтрации и преобразования данных

PolnySkvorets commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -30,6 +30,8 @@ public List<User> findAll() {
 
     @PostMapping
     public User create(@RequestBody User user) {
+        if (Optional.ofNullable(user).isEmpty())
```
Здесь можно проще заменить на user == null

PolnySkvorets reviewed (changes requested):
Посмотри одно замечание. Не нужно лишний раз использовать Optional, тут подойдёт обычное сравнение с null

MarchenkoDmitrii committed:
1. Исправленна проверка на пустое тело в запросе UserController

PolnySkvorets reviewed (approved):
Всё отлично!
